src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp

```diff
@@ -21,6 +21,8 @@
 
 #include <QSettings>
 
+QString QgsValueRelationFieldFormatter::FORM_SCOPE_FUNCTIONS_RE = QStringLiteral( "%1\\s*\\(\\s*'([^']+)'\\s*\\)" );
```

Comment:
What is the AST?

Reply:
The *Abstract Syntax Tree*, i.e. visiting the nodes instead of parsing the plain expression.

Reply:
> Can the AST of the expression nodes be used instead of a regex here?

Yes, I think it can be used (I need to learn how but this is another story).

> if I understand it correctly, this won't allow using layer variables or expression results?
No: the regexp is a template and gets filled with the function names that require a form scope to answer to a simple question: does this expression contain any function that requires a form scope?

I don't see the advantages in using AST over a simple regexp but maybe you can show me the light and give me a little tip about how to do this without writing thousands of lines of code. :smiley: 

I didn't find a tree traversal API in `QgsExpressionNode`: do I really need to try to cast every node starting from the root and branch for every node type? Is there a way to retrieve the list of all **functions** in the tree as there is for **variables**?

Reply:
Regex for parsing is not very tolerant `current_value( @main_field )` will probably not work. Plus one needs to take care of any current and upcoming syntax details (inline comments, escape sequences, quoting, ... ) to make it failproof.

You could probably add a method similar to `referencedColumns` and `referencedVariables` to get all function nodes (or even a template function to generically get all nodes by class type, similar to `QList<T> QObject::findChildren()` if you want to get crazy ;) ). Or as you say have some fun with traversing from outside and have some casting fun.

You'll then probably want to use `eval()` on the child param node to get the field name.

Reply:
+1 to adding a referencedFunctions method - should be just copy and paste from the existing methods

Reply:
@m-kuhn this is done in my last commit, but it does not solve the problem: once I get the names of the `referencedFunctions` how do I get the (evaluated) param values?

Do I also need a `QList<QgsExpressionFunction> referencedFunctionsObjects() const;` ?

Reply:
I guess you'll need a `QList<QgsExpressionNodeFunction> functionNodes( const QString &functionName = QString() /* if this is a null string return all functions */ )` or you'll get the function definition but no way to get to an evaluated parameter value.